src/NuGet.Core/NuGet.ProjectModel/HashWriter.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using NuGet.Common;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// Generates a hash from an object graph.
+    /// 
+    /// This is non-private only to facilitate unit testing.
+    /// </summary>
+    public sealed class HashWriter : IObjectWriter, IDisposable
+    {
+        private const int _defaultBufferSize = 4096;
+
+        // These delimiters are important in preserving the structure of an object map.
+        // Although the values were chosen out of familiarly with JSON, other values could be used.
+        private static readonly byte[] _null = new[] { (byte)0 };
```

Comment:
Why are we inventing a serialization format here? Can't we just use JSON directly? `JsonTextWriter` allows for incremental serialization. It's not immediately clear that these delimiters are sufficient or if escaping is necessary. This feels a bit like re-inventing the wheel.